stellargraph/core/graph.py

```diff
@@ -242,11 +242,25 @@ def _convert_from_node_data(data, node_type_map, node_types, dtype="f"):
     return data_index, data_arrays
 
 
-class StellarGraphBase:
+class StellarGraphFactory(type):
     """
-    StellarGraph class for undirected graph ML models. It stores both
-    graph information from a NetworkX Graph object as well as features
-    for machine learning.
+    Private class for instantiating the StellarGraph interface from
+    user-supplied information.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        if cls is StellarGraph:
+            return NetworkXStellarGraph(*args, is_directed=False, **kwargs)
+        elif cls is StellarDiGraph:
+            return NetworkXStellarGraph(*args, is_directed=True, **kwargs)
+        else:
+            return type.__call__(cls, *args, **kwargs)
+
+
+class StellarGraph(metaclass=StellarGraphFactory):
```

Comment:
I'm not convinced composition would buy us anything here. Firstly, every implemented method would have to delegate to the underlying implementation. Secondly, if the user instantiates StellarGraph with a NetworkX graph, then I don't see a particular problem if they get a NetworkXStellarGraph back.

Don't forget here that we still do not know exactly what methods the StellarGraph interface requires. This is being found by trial and error. Thus, the initial NetworkXStellarGraph keeps accessible implementation details that are not necessarily yet part of StellarGraph. I know this is a hack, but the point of the ticket was simply to make sure NetworkXStellarGraph no longer inherited from NetworkX graph types, without breaking existing functionality (where possible).

Reply:
As I have said before, there have been no firm requirements for this work, and we can treat this code as purely exploratory if we want. If the requirement is that users instantiating StellarGraph get StellarGraph, we can work to that, but it will mean a performance hit due to delegating all calls to the underlying implementation.

On the other hand, Java programmers don't seem to have a problem asking for a class implementing interface X and actually getting Y. Given that that `g=StellarGraph()` gives back something that satisfies `isinstance(g,StellarGraph)`, I'm not sure that casual users will notice, and expert users shouldn't be surprised.

The only odd thing about my implementation is that `g=StellarDiGraph()` satisfies `isinstance(g,StellarGraph)` by design, which might be a bit unexpected.

Reply:
It's true that it's not that unusual to have classes change, but the ramifications of this logic are more than just the types that get returned; e.g. the user has to be able to work out what arguments get passed to the `StellarGraph` constructor: documenting this usefully is much harder when there isn't an `__init__` that lists them all, and tooling like PyCharm & mypy likely doesn't work that great.

Reply:
This raises the question once again whether it is better to document the interface in the class or the `__init__`. In this case, the NetworkX-y interface is documented in the class, leaving the init free. Answers on the web suggest that people are more likely to type `help(StellarGraph)` than `help(StellarGraph.__init__)`, so class docs seem the best place.

Again, the problem we have is that we do not really have fully-defined arguments, due to organic growth of functionality. Multiply this by the polymorphism inherent in trying to support multiple interfaces, and I don't really see a neat encapsulation to be able to be explicit about all arguments.

Reply:
> This raises the question once again whether it is better to document the interface in the class or the __init__. In this case, the NetworkX-y interface is documented in the class, leaving the init free. Answers on the web suggest that people are more likely to type help(StellarGraph) than help(StellarGraph.__init__), so class docs seem the best place.

I think the location of the documentation is an orthogonal issue (also, `help(StellarGraph)` includes the documentation for `__init__`, and, in particular, its signature, which is what I care most about).

In particular, I'm getting at a more fundamental issue: having a computer-understandable definition of how a `StellarGraph` can be constructed. The best case for a library (and, indeed, anything) is no documentation needed, because the interface is "perfectly intuitive", meaning a user is guided down the right path automatically without having to read about how to use it ("the easy path is the right path"). Obviously, doing that entirely is unattainable for anything non-trivial, but it's still something that can be kept as a goal.

IME, a critical part of this is making it easy for computers to understand the easy path, because then the user's computer can tell the user when they've gone wrong, as soon as possible. In this specific case, this means having the expected arguments visible to the user and their editor/IDE/tooling. As this code is written, the only way to even know the names of the arguments is to read the documentation, which is something a computer cannot easily do.

Another part of it is making it easy for library maintainers to keep the documentation up to date, so when a user is forced to consult the docs, they are useful. With highly dynamic approaches, more documentation is required (i.e. more to maintain), and, there's less chance of tooling helping maintain it (e.g. if we remove or rename an argument, remind us to update the docs for it).

For instance, you may already know that IntelliJ/pycharm can be helpful about constructing things and calling functions:

![image](https://user-images.githubusercontent.com/1203825/69306252-fa1d7780-0c7a-11ea-9dd2-171b4a70d59a.png)

![image](https://user-images.githubusercontent.com/1203825/69306455-921b6100-0c7b-11ea-9301-6a21638d0ce4.png)

It can even show type annotations, if they exist on the function in question:

![image](https://user-images.githubusercontent.com/1203825/69306380-57b1c400-0c7b-11ea-9aef-9a82d7921dfd.png)

However, with the dynamic metaclasses, it can't show anything like this:

![image](https://user-images.githubusercontent.com/1203825/69306449-8def4380-0c7b-11ea-851e-d48d41afa65a.png)

There's no hint that I might want to be passing two arguments in there.

(NB. my example of IntelliJ/PyCharm is just a proxy for any tooling that might want to reason about our python code, other examples are mypy, pylint and even documentation generators.)

> Again, the problem we have is that we do not really have fully-defined arguments, due to organic growth of functionality. Multiply this by the polymorphism inherent in trying to support multiple interfaces, and I don't really see a neat encapsulation to be able to be explicit about all arguments.

This is a bit unfortunate, but it seems to me we should still think quite carefully about steps that make it harder to have great tooling.

Reply:
I understand that, and agree with everything you said. BUT in this case we still have the problem of having two incompatible class argument signatures, and will continue to have until major rationalisation of the NetworkX-based version can be done.

That said, the NodeData, EdgeData and EdgeCache interfaces could all be simplified back to standard init arguments.

Reply:
Moving forward, I think we can drop the NetworkX backend.  It was good for an interim solution in the previous PR #501 but I don't see it as useful moving forward.  This should make it straighforward to specify concrete class arguments.